### PR TITLE
Add support for precompilation

### PR DESCRIPTION
**Pull request by cocoa-xu**
This PR adds support for precompilation. (linking PR at [elixir-nx/stb_image#22](https://github.com/elixir-nx/stb_image/pull/22)).

I'll add more inline docs etc, but the code should be mostly okay for a first review. :)

### CODE REVIEW

**Comment by cocoa-xu** [comment]
Hi @josevalim, this should be ready for review now. :) 

I've added a precompilation guide for creating a precompiler module and using a precompiler. A demo precompiler module is available on [cocoa-xu/cc_precompiler](https://github.com/cocoa-xu/cc_precompiler), and an example project using this demo precompiler is available on [cocoa-xu/cc_precompiler_example](https://github.com/cocoa-xu/cc_precompiler_example).

I tried to make the precompilation document as easy to read and follow as possible, but please feel free to rephrase or change the wording in any part of that precompilation document directly. If there is anything I didn't make clear, please let me know, and I'll explain and have another try on that part.

**Comment by josevalim** [comment]
Thank you @cocoa-xu! I will merge this PR and do any improvements locally. I will ping back here once it is ready to go!

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

File: lib/mix/tasks/compile.make.ex [hunk 1]
```diff
@@ -99,69 +95,55 @@ defmodule Mix.Tasks.Compile.ElixirMake do
...
+      end
     end
+  end
+
+  defp precompiler_context(args, module) do
+    module = Module.concat([Mix.Tasks.ElixirMake, module])
 
-    @return
+    if Code.ensure_loaded?(module) do
+      Kernel.apply(module, :precompiler_context, [args])
```
Comment (josevalim) [inline lib/mix/tasks/compile.make.ex:143]: You declare this callback as optional, but here we always invoke it. Should we not make it optional?

File: lib/mix/tasks/elixir_make.precompile.ex [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+  @doc """
+  This callback will be invoked when the user executes the following commands:
+
+  - `mix elixir_make.fetch --only-local`
+  - `mix elixir_make.fetch --only-local --print`
+
+  The precompiler module should return the URL to a precompiled artefact of
+  the NIF library for current target (the "native" host).
+  """
+  @callback current_target_nif_url() :: String.t()
```
Comment (josevalim) [inline lib/mix/tasks/elixir_make.precompile.ex:89]: Should we have this function? Or should we retrieve the URL by calling `available_nif_urls()` and `current_target()`?

File: lib/mix/tasks/compile.make.ex (hunk 1 above)
Comment (josevalim) [reply lib/mix/tasks/compile.make.ex:143]: Actually we never directly use the context, so I have removed it. :)

**Pull request merged**